### PR TITLE
[v7] add ROOT base exception and RResult<T> class

### PR DESCRIPTION
**Pull request by jblomer**
The RException class is supposed to serve as base class for all ROOT
exceptions. It contains an `RError` member with diagnostic information.
The `RResult<T>` class can be used as a return value of operations that
may fail. The `RResult<T>` object wraps either a valid value or an `RError`.
If an error state remains unchecked, the `RResult` class will throw an
exception on destruction.

### CODE REVIEW

File: core/base/v7/inc/ROOT/RError.hxx [hunk 1]
```diff
@@ -0,0 +1,202 @@
+/// \file ROOT/RError.h
+/// \ingroup Base ROOT7
+/// \author Jakob Blomer <jblomer@cern.ch>
+/// \date 2019-12-11
+/// \warning This is part of the ROOT 7 prototype! It will change without notice. It might trigger earthquakes. Feedback
+/// is welcome!
+
+/*************************************************************************
+ * Copyright (C) 1995-2015, Rene Brun and Fons Rademakers.               *
```
Comment (pcanal) [inline core/base/v7/inc/ROOT/RError.hxx]:
```suggestion
 * Copyright (C) 1995-2019, Rene Brun and Fons Rademakers.               *
```

File: core/base/v7/inc/ROOT/RError.hxx [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+/**
+\class ROOT::Experimental::RException
+\ingroup Base
+\brief Base class for all ROOT issued exceptions
+*/
+// clang-format on
+class RException : public std::runtime_error {
+public:
+   explicit RException(const std::string &what)
+      : std::runtime_error(what + " [errno: " + std::to_string(errno) + "]") {}
```
Comment (pcanal) [inline core/base/v7/inc/ROOT/RError.hxx]: Isn't 'errno' only usefull for a subset of operation? (i.e. I/O)

File: core/base/v7/inc/ROOT/RError.hxx [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+};
+
+
+namespace Detail {
+
+// clang-format off
+/**
+\class ROOT::Experimental::Detail::RStatusType
+\ingroup Base
+\brief Values of this type have a distinct value that indicates an error, such as -1 for system calls
```
Comment (pcanal) [inline core/base/v7/inc/ROOT/RError.hxx]: You may want to call out to the fact that for the 'default' implementation 'all' values are consider an error.

File: core/base/v7/inc/ROOT/RError.hxx [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+};
+
+// clang-format off
+/**
+\class ROOT::Experimental::Detail::RStatus
+\ingroup Base
+\brief Wraps a return value such that unchecked error states trigger an exception
+
+The Status wrapper itself is movable but not copyable to prevent multiple exceptions from being thrown.
+The wrapper is also only stack allocatable because throwing an exception depends on the object going out of scope.
```
Comment (pcanal) [inline core/base/v7/inc/ROOT/RError.hxx]: Is that a 'programming rule' or is it enforced?

Comment (pcanal) [reply core/base/v7/inc/ROOT/RError.hxx]: nevermind I see the deletion of operation delete;

File: core/base/v7/inc/ROOT/RError.hxx [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+#else
+         if (!std::uncaught_exception())
+#endif
+         {
+            throw RException("unchecked error");
+         }
+      }
+   }
+
+   void MayThrow(const std::string &why)
```
Comment (pcanal) [inline core/base/v7/inc/ROOT/RError.hxx]: The name lead to think that this routine would most often throw (that is contradicted by the R__unlikely :) ).
What is the semantic/use-case of the function?

File: core/base/v7/test/base_exception.cxx [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+TEST(Exception, InstantExceptions)
+{
+   ROOT::Experimental::SetThrowInstantExceptions(true); // the default
+   bool passedFailure = false;
+
+   try {
+      TestFailure();
+      passedFailure = true;
+   } catch (const RException&) {
+   }
```
Comment (pcanal) [inline core/base/v7/test/base_exception.cxx]: For 'fun' shouldn't we also add an assignment in the catch close to check it is run?

File: core/base/v7/inc/ROOT/RError.hxx [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+
+   bool IsError()
+   {
+      fIsChecked = true;
+      return fStatus.IsError();
+   }
+   bool IsValid() { return !IsError(); }
+   void ClearError() { fIsChecked = true; }
+
+   operator typename T::ValueType_t() const { return fStatus.Get(); }
```
Comment (pcanal) [inline core/base/v7/inc/ROOT/RError.hxx]: Should using this operator count as 'checking' the error status?

File: core/base/v7/test/base_exception.cxx [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+   EXPECT_TRUE(rv.IsError());
+   // No exception / crash when the scope closes
+}
+
+
+TEST(Exception, DoubleThrow)
+{
+   ROOT::Experimental::SetThrowInstantExceptions(false);
+   try {
+      auto rv = TestFailure();
```
Comment (pcanal) [inline core/base/v7/test/base_exception.cxx]: Is capturing the value (```auto rv = ```) enough to count as a check or are you testing the order of throw?

File: core/base/v7/test/base_exception.cxx [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+      throw ExceptionX("something else went wrong");
+   } catch (const ExceptionX&) {
+      // This will not catch RException, so that the program crashes if rv throws
```
Comment (pcanal) [inline core/base/v7/test/base_exception.cxx]: This test seems to be testing that the Status destructor is not executed 'thanks to' the other explicit throw.  If this is the case, it might be beneficial to be more explicit in comment about the goal of the test.

File: core/base/v7/inc/ROOT/RError.hxx (hunk 2 above)
Comment (jblomer) [reply core/base/v7/inc/ROOT/RError.hxx]: That's right, actually we don't need to concern ourselves with `errno` at this point.

File: core/base/v7/inc/ROOT/RError.hxx (hunk 3 above)
Comment (jblomer) [reply core/base/v7/inc/ROOT/RError.hxx]: Changed to curiously recurring template pattern.

File: core/base/v7/inc/ROOT/RError.hxx (hunk 5 above)
Comment (jblomer) [reply core/base/v7/inc/ROOT/RError.hxx]: `MayThrow` is supposed to transition into the error state. If instant exceptions are turned on, an exception will be thrown right away.  Otherwise, the RStatus object expects to be checked and failure to do so will throw an exception on destruction.

File: core/base/v7/inc/ROOT/RError.hxx (hunk 7 above)
Comment (jblomer) [reply core/base/v7/inc/ROOT/RError.hxx]: The operator doesn't check but is currently only supposed to allow for using the RStatus object in expressions as if it was the wrapped type.  I thought it would make sense to require users of RStatus to explicitly spell out "checking" by calling either of `IsError()`, `IsValid()`, `ClearError()`.

File: core/base/v7/test/base_exception.cxx (hunk 9 above)
Comment (jblomer) [reply core/base/v7/test/base_exception.cxx]: Yup, done.

**Comment by Axel-Naumann** [comment]
FYI, the similar [`llvm::Error`](https://llvm.org/doxygen/classllvm_1_1Error.html)

[`llvm::Expected`](https://llvm.org/doxygen/classllvm_1_1Expected.html)
is basically a `std::variant<value, llvm::Error>`, where getting the value is only permitted after having checked the error state.

**Comment by Axel-Naumann** [comment]
I am not convinced yet by the immediate exception state. For discussion purposes, let's assume the following stack causes an error in `linux_read_from_network`, which is captured in a `RStatus` in `ROOT::ReadFromNetwork()`:
```
linux_read_from_network()
ROOT::ReadFromNetwork()
ROOT::RNetFile::Read()
ROOT::RNTupleIterator::operator++()
UserCode::modelBackground()
```

The immediate exception state would allow us to throw an exception in `ROOT::ReadFromNetwork()`. But `ROOT::RNetFile::Read()` might want/need to handle the error state, e.g. by retrying, or by accessing alternative sources. It might even be that a partial read is modeled as an error. This would require `ROOT::ReadFromNetwork()` to force the "immediate exception" state to a certain value and reset it later - which is error prone and awkward; it shows that we use a global state that needs local knowledge.

If "immediate exception" is false and `ROOT::ReadFromNetwork()` captures the error and returns it, same for all subsequent frames but `ROOT::RNTupleIterator::operator++()` (which cannot return a `RStatus`), then only `ROOT::RNTupleIterator::operator++()` will throw. This would shorten the stack trace shown by a potential crash to
```
ROOT::RNTupleIterator::operator++()
UserCode::modelBackground()
```
- we would not know the stack trace that actually triggered the exception. We would likely have some info captured in the error message (e.g. "ROOT::ReadFromNetwork() failed to call linux_read_from_network()"), but we'd miss the connection, i.e. `ROOT::RNetFile::Read()`. One way of handling that is by injecting state:
```
RStatus ROOT::RNetFile::Read() {
   RStatus res(ROOT::ReadFromNetwork());
   ...
   return R__STATUS_PASSTHROUGH(res); // registers res to have passed through "ROOT::RNetFile::Read()" by emplacing a frame string *if* in error state.
```
That's similar to [`llvm::PrettyStackTraceEntry`](https://llvm.org/doxygen/classllvm_1_1PrettyStackTraceEntry.html), see usage [here](https://github.com/llvm/llvm-project/blob/df494f7512b0ecebdf3d7be97695a1b6278c0336/clang/lib/CodeGen/BackendUtil.cpp#L909)

FYI, there has been a multi-hundred email discussion on exactly this topic on the C++ committee mailing list, with pretty much two camps: allow error code handling (with potential exceptions) versus just always throw in case of errors, and if that's slow, go fix the compiler / runtime library implementations (but ABI break).

**TL;DR**:
I agree that exceptions instead of error codes allow for nicer code. I don't think that we can have that nice code when we introduce `RStatus`, because the developer of that code must not assume a state for "immediate exceptions". In the end, we cannot use exceptions (across frames returning `RStatus`) and need to handle the error state, *always*. That leaves missing stack frames versus manipulation of a global (with developers dangerously and tacitly assuming a default state) - where IMHO it's more robust to have a mechanism to supply the missing stack frames (which can then - even better! - include user-facing diagnostic messages rather than just function names).

**Comment by Axel-Naumann** [comment]
I am uncertain about the missing separation of "error code" and "payload value" for some value types. E.g. `RStatus<FILE*>` versus `RStatus<std::string>` (where the string is read from a file) seem similar: the value represents the result of a potentially failing operation. But in the latter case the value does not encode failure, while in the former it does (the `FILE*` being `== nullptr`). Maybe that's just me, but maybe that calls for something similar to `llvm::Expected` that then handles the error case internally. Especially for pointers that might be a better interface, forcing a validity check.

**Comment by jblomer** [comment]
I think I can follow.  That would mean that we accept the cost of allocating strings (stack frame, error message) in the RStatus class in case of errors.

File: core/base/v7/inc/ROOT/RError.hxx (hunk 7 above)
Comment (pcanal) [reply core/base/v7/inc/ROOT/RError.hxx]: I don't understand your problem.  It seems with the current implementation the following will *unexpectedly* throw:
```
if (TestSucess(...)) {
   // go on
} else {
   // Try alternative path
}
```
or am I missing something?

Comment (jblomer) [reply core/base/v7/inc/ROOT/RError.hxx]: When `RStatus` captures a boolean, the correct code is admittedly not super intuitive:

    auto rv = TestSucess(...);
    if (rv) {
       rv.ClearError();
       // go on
    } else {
       // alternative
    }

What I wanted to avoid, however, is that just reading the wrapped value counts as checking, such as in

    int filedes = Open("/no/such/file"); // The returned RStatus decays into an int, has been automatically "checked" but filedes contains -1

Comment (pcanal) [reply core/base/v7/inc/ROOT/RError.hxx]: In your last example, the compiler will (sometimes :( ) complains that the value is assigned but not used ...

Comment (pcanal) [reply core/base/v7/inc/ROOT/RError.hxx]: humm ... it sounds like the pattern this code recommends is something like:
```
auto result = Open("/no/such/file");
if (result.IsValid()) {
   int filedesc = result.Get();
   // go on
} else {
   // print error message.
}
```
if that is the case, do we really need the conversion operator to int?  Can we instead have just a operator bool().  Then the 'only' broken case would be:
```
bool result = Open("/no/such/file");
// don't use result at all ...
```

**Comment by jblomer** [comment]
@Axel-Naumann @pcanal Reworked a bit along the discussed lines (see updated PR message).

File: core/base/v7/inc/ROOT/RError.hxx [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+
+   // Prevent heap construction of RResult objects
+   void *operator new(std::size_t size) = delete;
+   void *operator new(std::size_t, void *) = delete;
+   void *operator new[](std::size_t) = delete;
+   void *operator new[](std::size_t, void *) = delete;
+};
+
+/// Short-hand to return an RResult<T> in an error state
+#define R__FAIL(msg) return std::make_unique<ROOT::Experimental::RError>(msg, __func__, __FILE__, __LINE__)
```
Comment (Axel-Naumann) [inline core/base/v7/inc/ROOT/RError.hxx]: core/base/v7/inc/ROOT/RLogger.hxx uses
```
#if defined(_MSC_VER)
#define R__LOG_PRETTY_FUNCTION __FUNCSIG__
#else
#define R__LOG_PRETTY_FUNCTION __PRETTY_FUNCTION__
#endif

#define R__LOG_HERE(LEVEL, GROUP) \
   ROOT::Experimental::RLogEntry(LEVEL, GROUP).SetFile(__FILE__).SetLine(__LINE__).SetFunction(R__LOG_PRETTY_FUNCTION)
```
Maybe use that, for consistency?

Comment (Axel-Naumann) [inline core/base/v7/inc/ROOT/RError.hxx]: Do we need to hand `unique_ptr`s around, or can we return a `RError` object? Its move ctor should be just fine, so no actual copying should happen.

File: core/base/v7/inc/ROOT/RError.hxx [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+   void Throw() { throw RException(*fError); }
+
+   // Prevent heap construction of RResult objects
+   void *operator new(std::size_t size) = delete;
+   void *operator new(std::size_t, void *) = delete;
+   void *operator new[](std::size_t) = delete;
+   void *operator new[](std::size_t, void *) = delete;
+};
+
+/// Short-hand to return an RResult<T> in an error state
```
Comment (Axel-Naumann) [inline core/base/v7/inc/ROOT/RError.hxx]: This seems to return a `RError`, not a `RResult`?

File: core/base/v7/inc/ROOT/RError.hxx [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+   }
+   explicit operator bool() const
+   {
+      fIsChecked = true;
+      return !fError;
+   }
+   RError *GetError() { return fError.get(); }
+   void Throw() { throw RException(*fError); }
+
+   // Prevent heap construction of RResult objects
```
Comment (Axel-Naumann) [inline core/base/v7/inc/ROOT/RError.hxx]: Could you add an explanation why `RResult` must not be on the heap? (And I guess you know that you cannot really prevent that from happening anyway - `new optional<RResult>()` will have an `RResult` on the heap and it works just fine...

File: core/base/v7/inc/ROOT/RError.hxx (hunk 7 above)
Comment (pcanal) [reply core/base/v7/inc/ROOT/RError.hxx]: @jblomer Sorry.  I am not sure (i.e. I don't remember) how we settle that discussion.  Could you add a comment?

Comment (jblomer) [reply core/base/v7/inc/ROOT/RError.hxx]: It's now modified such that the `RResult` class's `bool` operator is used for checking for error/success. In order to get the encapsulated value, an explicit call the `RResult<T>::Get()` is required. There is now also support for `RResult<void>`, i.e. for functions that don't return anything but still can go wrong.

File: core/base/v7/inc/ROOT/RError.hxx (hunk 10 above)
Comment (jblomer) [reply core/base/v7/inc/ROOT/RError.hxx]: It's now using `R__LOG_PRETTY_FUNCTION`, should it use more of the RLogger infrastructure?

File: core/base/v7/inc/ROOT/RError.hxx (hunk 12 above)
Comment (jblomer) [reply core/base/v7/inc/ROOT/RError.hxx]: I clarified the comment.  It _should_ not be on the heap because than the destruction (and throwing of unchecked exceptions) can happen far away from the actual error.

File: core/base/v7/inc/ROOT/RError.hxx (hunk 11 above)
Comment (jblomer) [reply core/base/v7/inc/ROOT/RError.hxx]: Technically it's an `RError` but it gets converted into an `RResult<T>` because the constructor `RResult<T>(const RError &)` is not explicit.

File: core/base/v7/test/base_exception.cxx [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+{
+   bool exceptionThrown;
+
+   try {
+      exceptionThrown = false;
+      TestFailure();
+   } catch (const RException&) {
+      exceptionThrown = true;
+   }
+   EXPECT_TRUE(exceptionThrown);
```
Comment (hageboeck) [inline core/base/v7/test/base_exception.cxx]: gtest should allow to simplify this as follows:
```
EXPECT_THROW(TestFailure(), RException);
```

File: core/base/v7/test/base_exception.cxx [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+
+   try {
+      exceptionThrown = false;
+      TestSuccess();
+   } catch (const RException&) {
+      exceptionThrown = true;
+   }
+   EXPECT_FALSE(exceptionThrown);
+}
```
Comment (hageboeck) [inline core/base/v7/test/base_exception.cxx]:
```
EXPECT_NO_THROW(TestSuccess());
```

**Comment by pcanal** [comment]
@phsft-bot build with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by jblomer** [comment]
@phsft-bot build

File: core/base/v7/inc/ROOT/RError.hxx [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+   // the latest when leaving the scope. Heap allocated ill RResult objects can live much longer making it difficult
+   // to trace back the original failure.
+   void *operator new(std::size_t size) = delete;
+   void *operator new(std::size_t, void *) = delete;
+   void *operator new[](std::size_t) = delete;
+   void *operator new[](std::size_t, void *) = delete;
+};
+
+/// Short-hand to return an RResult<void> indicating success
+#define R__SUCCESS return ROOT::Experimental::RResult<void>();
```
Comment (Axel-Naumann) [inline core/base/v7/inc/ROOT/RError.hxx]: If it cannot go away, `return` should not be a part of it, such that an invocation reads:
```
return R__SUCCESS;
```
(But I'd really prefer to use a static constant.)

File: core/base/v7/test/base_exception.cxx [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+   bool exceptionThrown = false;
+   try {
+      TestChain(false);
+   } catch (const RException& e) {
+      exceptionThrown = true;
+      ASSERT_EQ(2U, e.GetError().GetStackTrace().size());
+      EXPECT_THAT(e.GetError().GetStackTrace().at(0).fFunction, ::testing::HasSubstr("TestSyscall(bool)"));
+      EXPECT_THAT(e.GetError().GetStackTrace().at(1).fFunction, ::testing::HasSubstr("TestChain(bool)"));
+   }
+   EXPECT_TRUE(exceptionThrown);
```
Comment (Axel-Naumann) [inline core/base/v7/test/base_exception.cxx]:
```suggestion
   try {
      TestChain(false);
      EXPECT_EQ("Above line should have thrown!", nullptr);
   } catch (const RException& e) {
      ASSERT_EQ(2U, e.GetError().GetStackTrace().size());
      EXPECT_THAT(e.GetError().GetStackTrace().at(0).fFunction, ::testing::HasSubstr("TestSyscall(bool)"));
      EXPECT_THAT(e.GetError().GetStackTrace().at(1).fFunction, ::testing::HasSubstr("TestChain(bool)"));
   }
```

File: core/base/v7/inc/ROOT/RError.hxx [hunk 17]
```diff
@@ -0,0 +1,230 @@
...
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT7_RError
+#define ROOT7_RError
+
+#include <ROOT/RConfig.hxx>
```
[the remark this answers is not in the document]
Comment (jblomer) [reply core/base/v7/inc/ROOT/RError.hxx]: clarified that it is for `R__unlikely`

File: core/base/v7/inc/ROOT/RError.hxx [hunk 18]
```diff
@@ -0,0 +1,230 @@
...
+\brief Captures diagnostics related to a ROOT runtime error
+*/
+// clang-format on
+class RError {
+public:
+   struct RLocation {
+      RLocation() = default;
+      RLocation(const std::string &func, const std::string &file, int line)
+         : fFunction(func), fSourceFile(file), fSourceLine(line) {}
+      std::string fFunction;
```
[the remark this answers is not in the document]
Comment (jblomer) [reply core/base/v7/inc/ROOT/RError.hxx]: changed to `const char *` and `fStackTrace` now reserves 32 frames on construction of `RLocation`.

File: core/base/v7/inc/ROOT/RError.hxx [hunk 19]
```diff
@@ -0,0 +1,230 @@
...
+unchecked, it will throw an exception.  RResult should only be allocated on the stack, which is helped by deleting the
+new operator.  RResult is movable but not copyable to avoid throwing exceptions multiple times.
+*/
+// clang-format on
+template <typename T>
+class RResult : public std::conditional_t<std::is_void<T>::value, Internal::RResultTypeVoid, Internal::RResultType<T>> {
+private:
+   /// This is the nullptr for an RResult representing success
+   std::unique_ptr<RError> fError;
+   /// This is safe because checking an RResult is not a multi-threaded operation
```
[the remark this answers is not in the document]
Comment (jblomer) [reply core/base/v7/inc/ROOT/RError.hxx]: One thing that doesn't work if the bool operator is non-const is in the unit tests

    auto res = Func();
    EXPECT_TRUE(res);

which made me believe there is other surprising behavior with a non-const bool operator.

File: core/base/v7/inc/ROOT/RError.hxx [hunk 20]
```diff
@@ -0,0 +1,230 @@
...
+   mutable bool fIsChecked{false};
+
+public:
+   /// Constructor is _not_ explicit in order to allow for `return T();` for functions returning RResult<T>
+   /// Only available if T is not void
+   template <typename Dummy = T, typename = typename std::enable_if_t<!std::is_void<T>::value, Dummy>>
+   RResult(const Dummy &value) : Internal::RResultType<T>(value) {}
```
[the remark this answers is not in the document]
Comment (jblomer) [reply core/base/v7/inc/ROOT/RError.hxx]: The problem is that then the compiler still complains about a `const void &` for `T == void`, even if it is eventually disabled.

File: core/base/v7/inc/ROOT/RError.hxx [hunk 21]
```diff
@@ -0,0 +1,230 @@
...
+   /// Constructor is _not_ explicit such that the RError returned by R__FAIL can be converted into an RResult<T>
+   /// for any T
+   RResult(const RError &error) : fError(std::make_unique<RError>(error)) {}
+
+   /// RResult<void> has a default contructor that creates an object representing success
+   template <typename Dummy = T, typename = typename std::enable_if_t<std::is_void<T>::value, Dummy>>
+   RResult() {}
```
[the remark this answers is not in the document]
Comment (jblomer) [reply core/base/v7/inc/ROOT/RError.hxx]: Instead of a `static const RResult` there is now a `static RResult Success()` for `T == void`, so that users can do `return RResult<void>::Success()` (instead of `R__SUCCESS`). Returning a `static const RResult kVoidSuccess` would have involved a copy and `RResult` is not copyable.

[the remark this answers is not in the document]
Comment (jblomer) [reply core/base/v7/inc/ROOT/RError.hxx]: Made private, called by `static RResult Success()`

File: core/base/v7/inc/ROOT/RError.hxx [hunk 22]
```diff
@@ -0,0 +1,230 @@
...
+
+   RResult(const RResult &other) = delete;
+   RResult(RResult &&other) = default;
+   RResult &operator =(const RResult &other) = delete;
+   RResult &operator =(RResult &&other) = default;
+
+   ~RResult() noexcept(false)
+   {
+      if (R__unlikely(fError && !fIsChecked)) {
+         // Prevent from throwing if the object is deconstructed in the course of stack unwinding for another exception
```
[the remark this answers is not in the document]
Comment (jblomer) [reply core/base/v7/inc/ROOT/RError.hxx]: It's reporting now the else case using `R__WARNING_HERE`

File: core/base/v7/inc/ROOT/RError.hxx [hunk 23]
```diff
@@ -0,0 +1,230 @@
...
+#endif
+         {
+            throw RException(*fError);
+         }
+      }
+   }
+
+   /// Only available if T is not void
+   template <typename Dummy = T>
+   typename std::enable_if_t<!std::is_void<T>::value, const Dummy &>
```
[the remark this answers is not in the document]
Comment (jblomer) [reply core/base/v7/inc/ROOT/RError.hxx]: The `Get()` method needs to have access to `fError` and `fIsChecked`, so pretty much to all of the constituents of `RResult`. An approach that avoids `enable_if` could simply be implemented by a  `RResult<void>` specialization (without any base class detour), which can be done of course but it would have some code duplication. What do you think?

File: core/base/v7/inc/ROOT/RError.hxx [hunk 24]
```diff
@@ -0,0 +1,230 @@
...
+
+   explicit operator bool() const
+   {
+      fIsChecked = true;
+      return !fError;
+   }
+
+   RError *GetError() { return fError.get(); }
+
+   void Throw() { throw RException(*fError); }
```
[the remark this answers is not in the document]
Comment (jblomer) [reply core/base/v7/inc/ROOT/RError.hxx]: Perhaps we can discuss this, I think I don't quite understand how or why `ThrowOutlined` would behave different  from the `Throw()` method.

File: core/base/v7/inc/ROOT/RError.hxx (hunk 15 above)
[the remark this answers is not in the document]
Comment (jblomer) [reply core/base/v7/inc/ROOT/RError.hxx]: Removed `R__SUCCESS` but with a slightly different approach.

File: core/base/v7/inc/ROOT/RError.hxx [hunk 25]
```diff
@@ -0,0 +1,230 @@
...
+   void *operator new[](std::size_t) = delete;
+   void *operator new[](std::size_t, void *) = delete;
+};
+
+/// Short-hand to return an RResult<void> indicating success
+#define R__SUCCESS return ROOT::Experimental::RResult<void>();
+/// Short-hand to return an RResult<T> in an error state; the RError is implicitly converted into RResult<T>
+#define R__FAIL(msg) return ROOT::Experimental::RError(msg, R__LOG_PRETTY_FUNCTION, __FILE__, __LINE__)
+/// Short-hand to return an RResult<T> value from a subroutine to the calling stack frame
+#define R__FORWARD_RESULT(res) if (res.GetError()) \
```
[the remark this answers is not in the document]
Comment (jblomer) [reply core/base/v7/inc/ROOT/RError.hxx]: Done using `static RResult &Forward(RResult &result, RError::RLocation &&sourceLocation)`

**Comment by jblomer** [comment]
@Axel-Naumann Many thanks!  Most of the comments are implemented, some issues might still need a clarification.

File: core/base/v7/inc/ROOT/RError.hxx (hunk 23 above)
[the remark this answers is not in the document]
Comment (jblomer) [reply core/base/v7/inc/ROOT/RError.hxx]: Actually, I just tried that: `RResult<void>` being a template specialization of `RResult<T>`.  Both `RResult<T>` and `RResult<void>` inherit from `Internal::RResultBase`, which implements common handling of the error case.  It looks to me nicer and simpler now and solves a few other issues because `enable_if` isn't required anymore.

A notable difference is that the `RResult<T>` data members now start with `fError` and `fChecked`, only then comes `fValue`.  I was wondering if we need to do anything specific for padding, but I'd believe the compiler does sensible padding based on the type of `T`...?

File: core/base/v7/inc/ROOT/RError.hxx (hunk 19 above)
[the remark this answers is not in the document]
Comment (Axel-Naumann) [reply core/base/v7/inc/ROOT/RError.hxx]: Thanks. How does that fail? I would expect it to work: https://www.godbolt.org/z/X_Rs2Y

File: core/base/v7/inc/ROOT/RError.hxx (hunk 23 above)
[the remark this answers is not in the document]
Comment (Axel-Naumann) [reply core/base/v7/inc/ROOT/RError.hxx]: Indeed, there should be no problem here. Plus all the offsets are statically known, so also perf-wise it's fine.

File: core/base/v7/inc/ROOT/RError.hxx (hunk 19 above)
[the remark this answers is not in the document]
Comment (Axel-Naumann) [reply core/base/v7/inc/ROOT/RError.hxx]: GTest failure is here: https://www.godbolt.org/z/uVJ_Yp

File: core/base/v7/inc/ROOT/RError.hxx (hunk 24 above)
[the remark this answers is not in the document]
Comment (Axel-Naumann) [reply core/base/v7/inc/ROOT/RError.hxx]: Clarified verbally as "user might compile user.cxx with `-fno-exceptions` and then they cannot `#include` RError.hxx as that contains `throw`. Instead, move `throw` into ROOT source file (i.e. "outlined" in contrast to "inlined")

File: core/base/v7/inc/ROOT/RError.hxx (hunk 19 above)
[the remark this answers is not in the document]
Comment (jblomer) [reply core/base/v7/inc/ROOT/RError.hxx]: Fixed in the unit test, no more use of mutable

**Comment by jblomer** [comment]
@Axel-Naumann It should be all done now

File: core/base/v7/inc/ROOT/RError.hxx (hunk 24 above)
[the remark this answers is not in the document]
Comment (pcanal) [reply core/base/v7/inc/ROOT/RError.hxx]: > and then they cannot #include RError.hxx

Do mean 'can not' as in "compilation will fail" or 'should not' as in "at run-time the code will do the 'wrong' thing?

It seems that if the user is compile the code with ```-fno-exceptions``` it would be 'surprising' if it throws exception, no?

**Comment by jblomer** [comment]
@Axel-Naumann Could you take another look at this?

File: core/base/v7/inc/ROOT/RError.hxx [hunk 26]
```diff
@@ -0,0 +1,260 @@
...
+      if (result.fError)
+         result.fError->AddFrame(std::move(sourceLocation));
+      return result;
+   }
+
+   explicit operator bool() { return Check(); }
+};
+
+/// Short-hand to return an RResult<T> in an error state; the RError is implicitly converted into RResult<T>
+#define R__FAIL(msg) return ROOT::Experimental::RError(msg, {R__LOG_PRETTY_FUNCTION, __FILE__, __LINE__})
```
Comment (Axel-Naumann) [inline core/base/v7/inc/ROOT/RError.hxx:254]: Why `return`?
